### DISCUSSION TITLE
Why is on_before_optimizer_step incompatible with accumulate_grad_batches ?

### DISCUSSION THREAD

**Discussion by NathanGodey** [Lightning-AI/pytorch-lightning#11331] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
The documentation about the `on_before_optimizer_step` hook mentions that:

> The hook is only called if gradients do not need to be accumulated.

However, in theory, combining both seems possible to me. 
I took a look at PR #8048 which is the implementation of this hook but it didn't seem clear there.

**Comment by rohitgr7** [comment, marked as answer]
hey @NathanGodey !

it doesn't mean that it won't be called if `accumulate_grad_batches > 1`, but only be called when accumulation is done and it's time to update the gradients with `optimizer.step`.

**Reply by NathanGodey** [reply]
Okay thanks! That is what I figured reading the PR. Don't you think the phrasing of the docs is misleading ?
Maybe : "The hook is called once the gradients have been accumulated" would make more sense ?

**Reply by rohitgr7** [reply]
feel free to send a PR with an update :)
